Stop VP from assuming that arrays are limited to INT32_MAX bytes

jdmpapin opened the pull request:
...and therefore at most `INT32_MAX` / `elementSize` elements. This could result in incorrect transformations, e.g. incorrect comparison folding.

These are (hopefully all of the) additional instances of the bug fixed in #7461.

Furthermore, there are two clean-up commits that delete dead code containing what would otherwise be additional instances of this bug:
- Delete unused signature-based overload of `VPArrayInfo::create()`.
- Clean up code in `constrainBndChkWithSpineChk()`

Review resulted in one more commit:
- VP: Propagate index lower bound to array length for `BNDCHKwithSpineCHK`

jdmpapin commented:
@hzongaro, would you mind reviewing?

hzongaro commented on compiler/optimizer/VPHandlers.cpp:
```diff
@@ -12279,64 +12251,14 @@ TR::Node *constrainBndChkWithSpineChk(OMR::ValuePropagation *vp, TR::Node *node)
       {
       TR::Node *objectRef = sizeNode->getFirstChild();
-      vp->addBlockConstraint(objectRef, TR::VPArrayInfo::create(vp, low, high, 0));
+      vp->addBlockConstraint(objectRef, TR::VPArrayInfo::create(vp, low, high + 1, 0));
```
Could this block constraint take advantage of any constraints on the `indexNode` to further narrow the range of the array size for `objectRef`?  I notice that code in [`constrainArrayCopyBndChk`](https://github.com/eclipse-omr/omr/blob/7dbf371f9571ad8bc6326f3a78a386772e0fcccd/compiler/optimizer/VPHandlers.cpp#L12353-L12360) does that.

jdmpapin replied:
I added a new commit to do this

hzongaro reviewed (commented):
Thanks for tracking these down!  The changes look correct.  I just had a couple of suggestions.

jdmpapin commented on compiler/optimizer/VPHandlers.cpp:
```diff
@@ -12207,70 +12207,42 @@ TR::Node *constrainBndChkWithSpineChk(OMR::ValuePropagation *vp, TR::Node *node)
-            constraint = TR::VPIntRange::create(vp, 0, upperBoundLimit-1);
-            foundArrayInfo = true;
-            }
+         int32_t lowerBoundLimit = arrayInfo->lowBound();
```
Removed

hzongaro reviewed (approved):
Changes look good.  Thanks!

Let me know whether you intended to squash any commits before I run testing.  I'm good with them as they are.

jdmpapin commented:
No, I wasn't meaning to squash

hzongaro commented:
Jenkins build all

hzongaro commented:
It looks like the [Windows build](https://ci.eclipse.org/omr/job/PullRequest-win_x86-64/4301/) failed due to warnings for `elementSize` values of type `uint32_t` that were passed to `om.maxArraySizeInElements`, which expects an `int32_t`.

```
15:28:16      31>c:\omr\workspace\pullrequest-win_x86-64\build\compiler\optimizer\vphandlers.cpp(12217): error C2220: warning treated as error - no 'object' file generated [C:\omr\workspace\PullRequest-win_x86-64\Build\build\jitbuilder\jitbuilder.vcxproj]
15:28:16      31>c:\omr\workspace\pullrequest-win_x86-64\build\compiler\optimizer\vphandlers.cpp(12217): warning C4244: 'initializing': conversion from 'int64_t' to 'int32_t', possible loss of data [C:\omr\workspace\PullRequest-win_x86-64\Build\build\jitbuilder\jitbuilder.vcxproj]
15:28:16      31>c:\omr\workspace\pullrequest-win_x86-64\build\compiler\optimizer\vphandlers.cpp(12318): warning C4244: 'initializing': conversion from 'int64_t' to 'int32_t', possible loss of data [C:\omr\workspace\PullRequest-win_x86-64\Build\build\jitbuilder\jitbuilder.vcxproj]
```

jdmpapin commented:
Looks like it was actually complaining about truncating the result of `maxArraySizeInElements()` to 32- from 64-bit. I don't know why it cares about these implicit truncations but not others in the same file, e.g. [this one](https://github.com/eclipse-omr/omr/blob/d944ab45a224dd23e1299b4625cc61e98ee732a7/compiler/optimizer/VPHandlers.cpp#L1959) from #7461

I've just introduced a cast, which is what happens - explicitly or implicitly - elsewhere in VP. It won't be correct if `maxArraySizeInElements()` returns something above 32-bit signed range in the future, but it's no worse than the existing code, and we're generally unprepared to deal with the possibility of that many elements anyway. For example, `BNDCHK` is 32-bit signed, and so is `arraylength`, and so are the array length bounds in `VPArrayInfo`.

Even though it wasn't causing the warning, I also changed `elementSize` to `int32_t` from `uint32_t` in the functions where the warnings happened. Its value is either the initial 1 or the result of `Node::getArrayStride()`, which is `int32_t`

hzongaro commented:
Jenkins build all

hzongaro commented:
[riscv build](https://ci.eclipse.org/omr/job/PullRequest-linux_riscv64/154/) failed because that build system is offline.  Merging.
